src/bptl/work_units/valid_sign/tasks.py

```diff
@@ -0,0 +1,382 @@
+                    raise NoAuth(
+                        f"Expected 'jwt' variable for {default_service.alias} in process variables"
+                    )
+                client.set_auth_value(jwt)
+                clients.append(client)
+
+            self._document_clients = clients
+        return self._document_clients
+
+    @staticmethod
+    def get_document_client(url: str, candidates: List[ZGWClient]) -> ZGWClient:
+        for candidate in candidates:
+            if url.startswith(candidate.base_url):
+                client = candidate
+                break
+        else:
+            raise DoesNotExist(f"No service found for url '{url}'")
+
+        return client
+
+    def format_signers(self, signers: List[dict]) -> List[dict]:
+        """Format the signer information into an array of JSON objects as needed by ValidSign."""
+
+        return [{"type": "SIGNER", "signers": [signer]} for signer in signers]
+
+    def _get_documents_from_api(self) -> List[Tuple[str, bytes]]:
+        """Retrieve the documents and their content from the Documenten API."""
+
+        logger.debug("Retrieving documents from Documenten API")
+
+        variables = self.task.get_variables()
+        document_urls = check_variable(variables, "documents")
+        # Building the document clients for all DRC services
+        document_clients = self.build_document_clients()
+
+        documents = []
+
+        current_total_documents_size = 0
+        for document_url in document_urls:
+            # Getting the appropriate client
+            document_client = self.get_document_client(document_url, document_clients)
+            # Retrieving the document
+            document_data = document_client.retrieve(
+                resource="enkelvoudiginformatieobject", url=document_url,
+            )
+
+            # Retrieving the content of the document
+            # Need use requests directly instead of `document_client.request()` since the response is not in JSON format
+            response = requests.get(
+                document_data["inhoud"],
+                headers=document_client.auth_header,
+                stream=True,
+            )
+
+            # Get the document size in Mb
+            document_size = int(response.headers["content-length"]) / 1e6
+
+            # If the size of the document is above the max size or if all the documents together have already reached
+            # the maximum size, write the file content to a temporary file
+            if (
+                document_size > settings.MAX_DOCUMENT_SIZE
+                or (current_total_documents_size + document_size)
+                > settings.MAX_TOT_DOCUMENT_SIZE
+            ):
+                tmp_file_object = TemporaryUploadedFile(
+                    name=f"{document_data['titel']}-{get_random_string(length=5)}.tempfile",
+                    content_type="application/octet-stream",
+                    size=sys.getsizeof(response.content),
```

Comment:
accessing `response.content` causes the entire document to be loaded into memory and is precisely what we don't want to do to prevent out-of-memory errors.

`document_data['bestandsomvang']` has the filesize in bytes, per the API docs.

Reply:
:facepalm: